party_time/party_time.lua

```diff
@@ -0,0 +1,345 @@
+local os = require("os")
+local ui = require("ui")
+local string = require("string")
+local command = require("command")
+local packets = require("packets")
+local settings = require("settings")
+local treasure = require("treasure")
+
+local pt_settings = {
+    ui = {
+        x = 145,
+        y = 440,
+        enabled = true,
+    },
+    auto = {
+        accept_invites = true,   -- also for auto sending invites upon party requests
+        decline_invites = false, -- also ignores party requests from players on blacklists
+    },
+    blacklist = {}, 
+    whitelist = {},
+}
+
+pt_settings = settings.load(pt_settings, 'settings.lua')
+
+local invite_dialog = {}
+local invite_dialog_state = {
+    title = 'Party Invite',
+    style = 'normal',
+    x = pt_settings.ui.x,
+    y = pt_settings.ui.y,
+    width = 179,
+    height = 96,
+    resizable = false,
+    moveable = true,
+    closable = true,
+}
+local unhandled_requests = {}
+local unhandeled_invite = false
+
+
+local alias = {
+    add = {
+        ['add'] = true,
+        ['a'] = true,
+        ['+'] = true,
+    },
+    remove = {
+        ['remove'] = true,
+        ['rm'] = true,
+        ['r'] = true,
+        ['-'] = true,
+    },
+    on = {
+        ['true'] = true,
+        ['yes'] = true,
+        ['on'] = true,
+        ['y'] = true,
+    },
+    off = {
+        ['false'] = true,
+        ['off'] = true,
+        ['no'] = true,
+        ['n'] = true,
+    }
+}
+
+-- Command Arg Types
+-- Create and Register types
+command.arg.register('names', '<names:string(%a+)>*')
+
+command.arg.register_type('boolean', {
+    check = function(str)
+        if alias.on[str] then
+            return true
+        elseif alias.off[str] then
+            return false
+        end
+
+        error('Expected a boolean value.')
+    end
+})
```

Comment:
Forgive me, but this post seems to have (d)evolved into a bit of a tutorial on the command lib :D @zetaepyon Feel free to wikify that if you're super bored.

Similar to this you could also register another type to help with differentiating different sub-commands:
```lua
command.arg.register_type('addremove', {
    check = function(str)
        if alias.add[str] then
            return 'add'
        elseif alias.remove[str] then
            return 'remove'
        end

        error('Expected bla')
    end,
})
```

That would later to allow just checking with against `'add'` and `'remove'`. Which would not immediately gain you much, but you can generalize it further:
```lua
local arg_lookups = {
    ['add'] = 'add',
    ['a'] = 'add',
    ['+'] = 'add',
    ['remove'] = 'remove',
    ['r'] = 'remove',
    ['-'] = 'remove',
    ['yes'] = true,
    ['y'] = true,
    ['on'] = true,
    ['no'] = false,
    ['n'] = false,
    ['off'] = false,
}

command.arg.register('lookup', {
    check = function(str)
        local res = lookup[str]
        if res == nil then
            error('bla')
        end

        return res
    end,
}
```

Now this would save a bit of code later on and you can always assume people entered the full command. However, now cross-contamination between the different value types would be possible. So you can pass in `yes` when it expects `whitelist` as an argument. Could either split the two up into different tables (which is what I'd do in this case), but just as a demonstration of another command lib feature, you can also add options to the type:
```lua
-- Assuming the same `lookups` table as above

command.arg.register_type('lookup', {
    -- "options" refer to the things passed to an argument definition in parentheses, like `string(%a+)`
    -- It's treated as a comma-separated list of values and they are passed as function arguments
    parse_options = function(...)
        -- With this function we create a table of values which we will check against
        local values = {...}
        for i = 1, select('#', ...) do
            values[select(i, ...)] = true
        end
        return values
    end,
    check = function(str, options)
        local value = arg_lookups[str]
        if value == nil or not options[tostring(value)] then
            error('Expected one of \'' .. table.concat(values) .. '\', received \'' .. str .. '\'.')
        end

        return value
    end,
    -- This function is used to generate the help text on invalid commands
    generate_name = function(options)
        return 'one_of'
    end,
    -- Same, this function returns the arguments
    generate_args = function(options)
        local allowed = {}
        for key, value in pairs(arg_lookups) do
            if options[tostring(value)] then
                allowed[#allowed + 1] = key
            end
        end

        return table.concat(allowed, ',')
    end,
})
```

Now you can do `/pt blacklist a` and it'll translate to `'add'`, so you can just check against that in the blacklist command handler. Similarly for `remove`.

Additionally, if you enter a wrong command you'll get this help text:
```
> Invalid command for '/pt blacklist'.
/pt blacklist <foo: one_of(-,+,remove,r,add,a)> <name: string(%a+)>*
```

Reply:
addressed however errors incorrect arg errors are not displaying properly as defined in the type that had been registered. 
Example:
```
Invalid command syntax for '/pt'.
/pt w <add_remove: lookup_add_remove> <name: string(%a+)>*
```